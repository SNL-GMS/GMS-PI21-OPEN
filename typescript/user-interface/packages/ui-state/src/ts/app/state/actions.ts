import { createAction } from '@reduxjs/toolkit';

/**
 * Action used to reset the application state.
 */
export const reset = createAction('reset');
