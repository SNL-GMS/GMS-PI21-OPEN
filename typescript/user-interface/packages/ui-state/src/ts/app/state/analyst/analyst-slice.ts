/* eslint-disable @typescript-eslint/no-magic-numbers */
import type { WaveformTypes } from '@gms/common-model';
import { CommonTypes } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND } from '@gms/common-util';
import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import type { AnalystState } from './types';
import { AlignWaveformsOn, WaveformDisplayMode, WaveformSortType } from './types';

/**
 * The initial state for the analyst state.
 */
export const analystInitialState: AnalystState = {
  channelFilters: {},
  defaultSignalDetectionPhase: CommonTypes.PhaseType.P,
  effectiveNowTime: null,
  historyActionInProgress: 0,
  hotkeyCycleOverrides: {},
  location: {
    selectedLocationSolutionId: null,
    selectedLocationSolutionSetId: null,
    selectedPreferredLocationSolutionId: null,
    selectedPreferredLocationSolutionSetId: null
  },
  measurementMode: {
    entries: {},
    mode: WaveformDisplayMode.DEFAULT
  },
  openEventId: null,
  openLayoutName: null,
  sdIdsToShowFk: [],
  selectedEventIds: [],
  selectedFilterList: null,
  selectedFilterIndex: null,
  selectedSdIds: [],
  selectedSortType: WaveformSortType.stationNameAZ,
  eventListOpenEventTriggered: false,
  mapOpenEventTriggered: false,
  alignWaveformsOn: AlignWaveformsOn.TIME,
  phaseToAlignOn: null
};

/**
 * The analyst reducer slice.
 */
export const analystSlice = createSlice({
  name: 'analyst',
  initialState: analystInitialState,
  reducers: {
    /**
     * Decrements the history action is in progress
     *
     * @param state the state
     * @param action the action
     */
    decrementHistoryAction(state) {
      state.historyActionInProgress -= 1;
    },

    /**
     * Increments the history action is in progress
     *
     * @param state the state
     * @param action the action
     */
    incrementHistoryAction(state) {
      state.historyActionInProgress += 1;
    },

    /**
     * Sets the channel filters.
     * (selected waveform filter for a give channel id)
     *
     * @param state the state
     * @param action the action
     */
    setChannelFilters(state, action: PayloadAction<Record<string, WaveformTypes.WaveformFilter>>) {
      state.channelFilters = action.payload;
    },

    /**
     * Sets the signal detection default phase.
     * The selected phase type that will be used for the creation of
     * a new a signal detection.
     *
     * @param state the state
     * @param action the action
     */
    setDefaultSignalDetectionPhase(state, action: PayloadAction<CommonTypes.PhaseType>) {
      state.defaultSignalDetectionPhase = action.payload;
    },

    /**
     * Sets the effective now time. Sets it to Date.now if argument is not provided
     *
     * @param state the state
     * @param action the action
     */
    setEffectiveNowTime(state, action?: PayloadAction<number>) {
      state.effectiveNowTime = action.payload
        ? action.payload
        : Date.now() / MILLISECONDS_IN_SECOND;
    },

    /**
     * Set the current filter list which defines the filters available.
     */
    setSelectedFilterList(state, action: PayloadAction<string>) {
      state.selectedFilterList = action.payload;
    },

    /**
     * Set the current filter list which defines the filters available
     */
    setSelectedFilterIndex(state, action: PayloadAction<number | null>) {
      state.selectedFilterIndex = action.payload;
    },

    /**
     * Sets a single override to the default hotkey state for the filter at the provided index
     */
    setIsFilterWithinHotkeyCycle(
      state,
      action: PayloadAction<{ index: number; isWithinCycle: boolean }>
    ) {
      if (!state.selectedFilterList) {
        throw new Error('cannot override hotkey cycle with no filter list selected');
      }
      if (!state.hotkeyCycleOverrides[state.selectedFilterList]) {
        state.hotkeyCycleOverrides[state.selectedFilterList] = {};
      }
      state.hotkeyCycleOverrides[state.selectedFilterList][action.payload.index] =
        action.payload.isWithinCycle;
    },

    /**
     * Sets overrides to the default hotkey state, such that each key is an index, and the boolean
     * value is the user supplied override.
     */
    setFilterHotkeyCycleOverridesForCurrentFilterList(
      state,
      action: PayloadAction<Record<number, boolean>>
    ) {
      state.hotkeyCycleOverrides[state.selectedFilterList] = action.payload;
    },

    /**
     * Sets the measurement mode entries.
     * Record (dictionary) of signal detection ids to boolean value indicating if the
     * amplitude measurement should be displayed (visible).
     *
     * @param state the state
     * @param action the action
     */
    setMeasurementModeEntries(state, action: PayloadAction<Record<string, boolean>>) {
      state.measurementMode.entries = action.payload;
    },

    /**
     * Sets the waveform display mode
     *
     * @param state the state
     * @param action the action
     */
    setMode(state, action: PayloadAction<WaveformDisplayMode>) {
      state.measurementMode.mode = action.payload;
    },

    /**
     * Sets the open event id.
     *
     * @param state the state
     * @param action the action
     */
    setOpenEventId(state, action: PayloadAction<string>) {
      state.openEventId = action.payload;
    },

    /**
     * Sets the open layout name.
     *
     * @param state the state
     * @param action the action
     */
    setOpenLayoutName(state, action: PayloadAction<string>) {
      state.openLayoutName = action.payload;
    },

    /**
     * Sets the signal detection ids that have been marked to show FK.
     *
     * @param state the state
     * @param action the action
     */
    setSdIdsToShowFk(state, action: PayloadAction<string[]>) {
      state.sdIdsToShowFk = action.payload;
    },

    /**
     * Sets the selected event ids.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedEventIds(state, action: PayloadAction<string[]>) {
      state.selectedEventIds = action.payload;
    },

    /**
     * Sets flag indicating an analyst attempted to open an event from the event list
     *
     * @param state the state
     * @param action the action
     */
    setEventListOpenTriggered(state, action: PayloadAction<boolean>) {
      state.eventListOpenEventTriggered = action.payload;
    },

    /**
     * Sets flag indicating an analyst attempted to open an event from the map
     *
     * @param state the state
     * @param action the action
     */
    setMapOpenTriggered(state, action: PayloadAction<boolean>) {
      state.mapOpenEventTriggered = action.payload;
    },

    /**
     * Sets the selected location solution id.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedLocationSolutionId(state, action: PayloadAction<string>) {
      state.location.selectedLocationSolutionId = action.payload;
    },

    /**
     * Sets the selected location solution set id.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedLocationSolutionSetId(state, action: PayloadAction<string>) {
      state.location.selectedLocationSolutionSetId = action.payload;
    },

    /**
     * Sets the  selected preferred location solution id.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedPreferredLocationSolutionId(state, action: PayloadAction<string>) {
      state.location.selectedPreferredLocationSolutionId = action.payload;
    },

    /**
     * Sets the selected preferred location solution set id.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedPreferredLocationSolutionSetId(state, action: PayloadAction<string>) {
      state.location.selectedPreferredLocationSolutionSetId = action.payload;
    },

    /**
     * Sets the selected signal detection ids.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedSdIds(state, action: PayloadAction<string[]>) {
      state.selectedSdIds = action.payload;
    },

    /**
     * Sets the selected sort type.
     *
     * @param state the state
     * @param action the action
     */
    setSelectedSortType(state, action: PayloadAction<WaveformSortType>) {
      state.selectedSortType = action.payload;
    },
    /**
     * Sets the selected alignment.
     *
     * @param state the state
     * @param action the action
     */
    setAlignWaveformsOn(state, action: PayloadAction<AlignWaveformsOn>) {
      state.alignWaveformsOn = action.payload;
    },
    /**
     * Sets the selected sort type.
     *
     * @param state the state
     * @param action the action
     */
    setPhaseToAlignOn(state, action: PayloadAction<CommonTypes.PhaseType>) {
      state.phaseToAlignOn = action.payload;
    }
  }
});

/**
 * The analyst actions.
 */
export const analystActions = analystSlice.actions;
