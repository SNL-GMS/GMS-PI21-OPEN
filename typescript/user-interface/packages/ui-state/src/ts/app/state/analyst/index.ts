export * from './analyst-slice';
export * as AnalystWorkspaceOperations from './operations';
export * from './selectors';
export * as AnalystWorkspaceTypes from './types';
