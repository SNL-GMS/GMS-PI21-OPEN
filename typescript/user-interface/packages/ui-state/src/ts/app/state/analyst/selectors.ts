import type { FilterListTypes } from '@gms/common-model';
import { createSelector } from '@reduxjs/toolkit';

import { selectFilterLists } from '../../api/signal-enhancement-configuration/selectors';
import type { AppState } from '../../store';
import type { HotkeyCycleList } from './types';

/**
 * @returns the selected filter list from the provided redux state
 */
export const selectSelectedFilterListName = (state: AppState): string =>
  state.app.analyst.selectedFilterList;

/**
 * A selector that derives the selected filter list from the signalEnhancementConfiguration query (filters)
 * and the currently selected filter list.
 *
 * @returns the selected filter list. Note that this is memoized by createSelector.
 * @see https://redux.js.org/usage/deriving-data-selectors
 */
export const selectSelectedFilterList: (
  state: AppState
) => FilterListTypes.FilterList = createSelector(
  [selectSelectedFilterListName, selectFilterLists],
  (selectedFLName, filterLists) =>
    filterLists != null ? filterLists.find(fl => fl.name === selectedFLName) : undefined
);

/**
 * a selector to get the selected filter out of the redux state.
 *
 * @example
 * const selectedFilterIndex = useAppSelector(selectSelectedFilterIndex);
 *
 * @param state the AppState
 * @returns the index of the selected filter, or null if there is no selection
 */
export const selectSelectedFilterIndex: (state: AppState) => number | null = state =>
  state.app.analyst.selectedFilterIndex;

/**
 * a selector to get the selected filter out of the redux state.
 *
 * @example
 * const selectedFilter = useAppSelector(selectSelectedFilter);
 *
 * @param state the AppState
 * @returns the index of the selected filter, or null if there is no selection
 */
export const selectSelectedFilter: (state: AppState) => FilterListTypes.Filter = createSelector(
  [selectSelectedFilterIndex, selectSelectedFilterList],
  (index: number, selectedFilterList: FilterListTypes.FilterList) => {
    return selectedFilterList[index];
  }
);

/**
 * @returns the record of records of user supplied overrides to the hotkey cycle, such that each key
 * is the index of a record of hotkeyCycle overrides
 * @see selectHotkeyCycle for an array of boolean values indicating which filters are in the cycle
 * @see selectHotkeyCycleOverrides for this FilterList's record of overrides
 */
export const selectAllHotkeyCycleOverrides: (
  state: AppState
) => Record<string, Record<number, boolean>> = state => state.app.analyst.hotkeyCycleOverrides;

/**
 * @returns the record of user supplied overrides to the hotkey cycle, such that each key
 * is the index of a filter, and it's boolean value represents the user-set overridden state.
 * This is intended primarily for internal use, and components that care about whether a filter
 * is within the hotkey cycle should use the @function selectHotkeyCycle.
 * @see selectHotkeyCycle for an array of boolean values indicating which filters are in the cycle
 * @see selectAllHotkeyCycleOverrides for a record of all overrides
 */
export const selectHotkeyCycleOverrides: (
  state: AppState
) => Record<number, boolean> = createSelector(
  selectSelectedFilterListName,
  selectAllHotkeyCycleOverrides,
  (filterListName, allHotkeyCycles) => allHotkeyCycles[filterListName]
);

/**
 * @returns a HotkeyCycleList that contains a boolean at the corresponding index of each filter within
 * the FilterList.filter ordered list.
 */
export const selectHotkeyCycle: (state: AppState) => HotkeyCycleList = createSelector(
  [selectSelectedFilterList, selectHotkeyCycleOverrides],
  (filterList: FilterListTypes.FilterList, hotkeyCycleOverrides: Record<number, boolean>) => {
    if (!filterList?.filters) {
      return [];
    }
    return filterList.filters.map((fl, index) =>
      hotkeyCycleOverrides != null && hotkeyCycleOverrides[index] != null
        ? hotkeyCycleOverrides[index]
        : fl.withinHotKeyCycle
    );
  }
);
