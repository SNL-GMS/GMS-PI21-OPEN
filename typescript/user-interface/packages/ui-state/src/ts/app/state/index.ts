export * as AppActions from './actions';
export * from './analyst';
export * from './common';
export * from './data-acquisition';
export * from './events';
export * from './keypress';
export * from './map';
export * as AppOperations from './operations';
export * from './signal-detections';
export * from './station-properties-configuration';
export * from './system-message';
export * from './user-session';
export * from './waveform';
export * from './workflow';
