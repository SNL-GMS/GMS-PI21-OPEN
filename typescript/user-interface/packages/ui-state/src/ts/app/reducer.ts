import { combineReducers } from '@reduxjs/toolkit';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { dataSlice } from './api/data/data-slice';
import { reducer as appReducer } from './state/reducer';

export const reducer = combineReducers({
  // application api (queries)
  [systemEventGatewayApiSlice.reducerPath]: systemEventGatewayApiSlice.reducer,
  [eventManagerApiSlice.reducerPath]: eventManagerApiSlice.reducer,
  [processingConfigurationApiSlice.reducerPath]: processingConfigurationApiSlice.reducer,
  [processingStationApiSlice.reducerPath]: processingStationApiSlice.reducer,
  [signalEnhancementConfigurationApiSlice.reducerPath]:
    signalEnhancementConfigurationApiSlice.reducer,
  [ssamControlApiSlice.reducerPath]: ssamControlApiSlice.reducer,
  [sohAceiApiSlice.reducerPath]: sohAceiApiSlice.reducer,
  [stationDefinitionSlice.reducerPath]: stationDefinitionSlice.reducer,
  [systemMessageDefinitionApiSlice.reducerPath]: systemMessageDefinitionApiSlice.reducer,
  [userManagerApiSlice.reducerPath]: userManagerApiSlice.reducer,
  [workflowApiSlice.reducerPath]: workflowApiSlice.reducer,
  // data state
  [dataSlice.name]: dataSlice.reducer,
  // application state
  app: appReducer
});
