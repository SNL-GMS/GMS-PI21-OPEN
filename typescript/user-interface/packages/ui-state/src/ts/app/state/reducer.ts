import type { CombinedState } from '@reduxjs/toolkit';
import { combineReducers, createReducer } from '@reduxjs/toolkit';
import type { ReducerWithInitialState } from '@reduxjs/toolkit/dist/createReducer';

import {
  analystInitialState,
  analystSlice,
  commonInitialState,
  commonSlice,
  dataAcquisitionInitialState,
  dataAcquisitionSlice,
  waveformInitialState,
  waveformSlice,
  workflowInitialState,
  workflowSlice
} from '.';
import { reset } from './actions';
import type { AnalystState } from './analyst/types';
import type { CommonState } from './common/types';
import type { DataAcquisitionState } from './data-acquisition/types';
import type { EventsState } from './events';
import { eventsInitialState, eventsSlice } from './events';
import type { MapState } from './map';
import { mapInitialState, mapSlice } from './map';
import type { SignalDetectionsState } from './signal-detections';
import { signalDetectionsInitialState, signalDetectionsSlice } from './signal-detections';
import {
  stationPropertiesConfigurationInitialState,
  stationPropertiesConfigurationSlice
} from './station-properties-configuration';
import type { StationPropertiesConfigurationState } from './station-properties-configuration/types';
import {
  systemMessageInitialState,
  systemMessageSlice
} from './system-message/system-message-slice';
import type { SystemMessageState } from './system-message/types';
import type { UserSessionState } from './user-session/types';
import { userSessionInitialState, userSessionSlice } from './user-session/user-session-slice';
import type { WaveformState } from './waveform/types';
import type { WorkflowState } from './workflow/types';

// combine all reducers
const reducers = combineReducers<{
  analyst: AnalystState;
  stationPropertiesConfiguration: StationPropertiesConfigurationState;
  common: CommonState;
  dataAcquisition: DataAcquisitionState;
  events: EventsState;
  map: MapState;
  signalDetections: SignalDetectionsState;
  systemMessage: SystemMessageState;
  userSession: UserSessionState;
  waveform: WaveformState;
  workflow: WorkflowState;
}>({
  [analystSlice.name]: analystSlice.reducer,
  [stationPropertiesConfigurationSlice.name]: stationPropertiesConfigurationSlice.reducer,
  [commonSlice.name]: commonSlice.reducer,
  [dataAcquisitionSlice.name]: dataAcquisitionSlice.reducer,
  [eventsSlice.name]: eventsSlice.reducer,
  [mapSlice.name]: mapSlice.reducer,
  [signalDetectionsSlice.name]: signalDetectionsSlice.reducer,
  [systemMessageSlice.name]: systemMessageSlice.reducer,
  [userSessionSlice.name]: userSessionSlice.reducer,
  [waveformSlice.name]: waveformSlice.reducer,
  [workflowSlice.name]: workflowSlice.reducer
});

// Infer the `State` type from the reducer itself
type State = ReturnType<typeof reducers>;

/**
 * The app initial state
 */
export const initialState: State = {
  [analystSlice.name]: analystInitialState,
  [stationPropertiesConfigurationSlice.name]: stationPropertiesConfigurationInitialState,
  [commonSlice.name]: commonInitialState,
  [dataAcquisitionSlice.name]: dataAcquisitionInitialState,
  [eventsSlice.name]: eventsInitialState,
  [mapSlice.name]: mapInitialState,
  [signalDetectionsSlice.name]: signalDetectionsInitialState,
  [systemMessageSlice.name]: systemMessageInitialState,
  [userSessionSlice.name]: userSessionInitialState,
  [waveformSlice.name]: waveformInitialState,
  [workflowSlice.name]: workflowInitialState
};

/**
 * The application state reducer
 *
 * Type inference here was causing errors (it was attempting to export types from third party packages).
 * This caused build errors, which is likely an issue with RTK, immer, or typescript itself. Explicitly
 * typing it here seems to make typescript happy.
 */
export const reducer: ReducerWithInitialState<CombinedState<{
  analyst: AnalystState;
  stationPropertiesConfiguration: StationPropertiesConfigurationState;
  common: CommonState;
  dataAcquisition: DataAcquisitionState;
  events: EventsState;
  map: MapState;
  signalDetections: SignalDetectionsState;
  systemMessage: SystemMessageState;
  userSession: UserSessionState;
  waveform: WaveformState;
  workflow: WorkflowState;
}>> = createReducer(initialState, builder => {
  builder
    .addCase(reset, state => {
      state.analyst = initialState.analyst;
      state.stationPropertiesConfiguration = initialState.stationPropertiesConfiguration;
      state.common = initialState.common;
      state.dataAcquisition = initialState.dataAcquisition;
      state.events = initialState.events;
      state.map = initialState.map;
      state.signalDetections = initialState.signalDetections;
      state.systemMessage = initialState.systemMessage;
      state.userSession = initialState.userSession;
      state.waveform = initialState.waveform;
      state.workflow = initialState.workflow;
    })
    .addDefaultCase(reducers);
});
