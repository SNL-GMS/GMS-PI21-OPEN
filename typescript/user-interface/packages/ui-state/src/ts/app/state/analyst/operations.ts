import type { LegacyEventTypes } from '@gms/common-model';
import { UILogger } from '@gms/ui-util';

import { selectPreferredFilterListForActivity } from '../../api/signal-enhancement-configuration/selectors';
import { signalEnhancementConfigurationApiSlice } from '../../api/signal-enhancement-configuration/signal-enhancement-api-slice';
import type { AppDispatch, AppState } from '../../store';
import { analystActions } from './analyst-slice';
import { selectSelectedFilterList, selectSelectedFilterListName } from './selectors';
import { WaveformDisplayMode, WaveformSortType } from './types';

const logger = UILogger.create('[UI State Manager - Analyst]', process.env.GMS_LOG_UI_STATE_STORE);

/**
 * Redux operation for setting the mode.
 *
 * @param mode the mode to set
 */
export const setMode = (mode: WaveformDisplayMode) => (dispatch: AppDispatch): void => {
  dispatch(analystActions.setMode(mode));
};

/**
 * Redux operation for setting the measurement mode entries.
 *
 * @param entries the measurement mode entries to set
 */
export const setMeasurementModeEntries = (entries: Record<string, boolean>) => (
  dispatch: AppDispatch
): void => {
  dispatch(analystActions.setMeasurementModeEntries(entries));
};

/**
 * Redux operation for setting the selected location solution.
 *
 * @param locationSolutionSetId the location solution set id
 * @param locationSolutionId the location solution id
 */
export const setSelectedLocationSolution = (
  locationSolutionSetId: string,
  locationSolutionId: string
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (getState().app.analyst.location.selectedLocationSolutionSetId !== locationSolutionSetId) {
    dispatch(analystActions.setSelectedLocationSolutionSetId(locationSolutionSetId));
  }

  if (getState().app.analyst.location.selectedLocationSolutionId !== locationSolutionId) {
    dispatch(analystActions.setSelectedLocationSolutionId(locationSolutionId));
  }
};

/**
 * Redux operation for setting the selected preferred location solution.
 *
 * @param preferredLocationSolutionSetId the preferred location solution set id
 * @param preferredLocationSolutionId the preferred location solution id
 */
export const setSelectedPreferredLocationSolution = (
  preferredLocationSolutionSetId: string,
  preferredLocationSolutionId: string
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (
    getState().app.analyst.location.selectedPreferredLocationSolutionSetId !==
    preferredLocationSolutionSetId
  ) {
    dispatch(
      analystActions.setSelectedPreferredLocationSolutionSetId(preferredLocationSolutionSetId)
    );
  }

  if (
    getState().app.analyst.location.selectedPreferredLocationSolutionId !==
    preferredLocationSolutionId
  ) {
    dispatch(analystActions.setSelectedPreferredLocationSolutionId(preferredLocationSolutionId));
  }
};

/**
 * Redux operation for setting the current open event id.
 *
 * @param event the event to set
 * @param latestLocationSolutionSet
 * @param preferredLocationSolutionId
 */
export const setOpenEventId = (
  event: LegacyEventTypes.Event | undefined,
  latestLocationSolutionSet: LegacyEventTypes.LocationSolutionSet | undefined,
  preferredLocationSolutionId: string | undefined
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (getState().app.workflow.timeRange && event) {
    if (getState().app.analyst.openEventId !== event.id) {
      dispatch(analystActions.setOpenEventId(event.id));
      dispatch(analystActions.setSelectedEventIds([event.id]));
      dispatch(analystActions.setSelectedSortType(WaveformSortType.distance));

      // set the default (latest) location solution
      setSelectedLocationSolution(
        latestLocationSolutionSet ? latestLocationSolutionSet.id : undefined,
        latestLocationSolutionSet ? latestLocationSolutionSet.locationSolutions[0].id : undefined
      )(dispatch, getState);

      // set the default (latest) preferred location solution
      setSelectedPreferredLocationSolution(
        latestLocationSolutionSet.id,
        preferredLocationSolutionId
      )(dispatch, getState);
    }
  } else {
    dispatch(analystActions.setOpenEventId(undefined));
    dispatch(analystActions.setSelectedEventIds([]));
    dispatch(analystActions.setSelectedSortType(WaveformSortType.stationNameAZ));
    dispatch(analystActions.setMeasurementModeEntries({}));
    // update the selected location and preferred location solutions
    setSelectedLocationSolution(undefined, undefined)(dispatch, getState);
    setSelectedPreferredLocationSolution(undefined, undefined)(dispatch, getState);
  }

  // TODO: confirm should we be clearing????
  dispatch(analystActions.setSelectedSdIds([]));
  dispatch(analystActions.setSdIdsToShowFk([]));
  setMode(WaveformDisplayMode.DEFAULT)(dispatch);
};

/**
 * creates an action that sets the selected filter list to be the preferred filter list from the list of
 * preferred filter lists returned by the filter definition query.
 * Logs warnings if it is being misused, for example, if the preferred filter list for the activity is nullish,
 * or if the selectedFilterList was already set.
 *
 * @see signalEnhancementConfigurationApiSlice
 */
export const setPreferredFilterList = () => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const state = getState();
  const preferredFilterListForActivity = selectPreferredFilterListForActivity(state);
  const selectedFilterList = selectSelectedFilterListName(state);
  if (selectedFilterList == null && preferredFilterListForActivity != null) {
    dispatch(analystActions.setSelectedFilterList(preferredFilterListForActivity));
  } else if (selectedFilterList != null && preferredFilterListForActivity == null) {
    logger.warn(`Cannot set preferred filter list because it is not defined.`);
  } else if (selectedFilterList == null && preferredFilterListForActivity == null) {
    logger.warn(
      `Cannot set preferred filter list because neither it nor the selected filter list is defined.`
    );
  } else {
    logger.warn(
      `Cannot set preferred filter list ${preferredFilterListForActivity} because ${selectedFilterList} was already selected.`
    );
  }
};

/**
 * Sets the default filter from the currently selected filter list.
 * Logs a warning if no selected filter is found.
 * This will subscribe the component to the results of the filter list query.
 * TODO: We need to unsubscribe? How? Where?
 */
export const setDefaultFilter = () => async (
  dispatch: AppDispatch,
  getState: () => AppState
): Promise<void> => {
  const state = getState();
  // subscribe to this state
  await dispatch(
    signalEnhancementConfigurationApiSlice.endpoints.getFilterListsDefinition.initiate()
  );
  const fl = selectSelectedFilterList(state);
  if (fl == null) {
    logger.warn('cannot set default filter list');
    return;
  }
  dispatch(analystActions.setSelectedFilterIndex(fl.defaultFilterIndex));
};
