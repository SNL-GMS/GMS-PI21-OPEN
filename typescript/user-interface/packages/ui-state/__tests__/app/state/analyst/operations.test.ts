import type * as Redux from 'redux';
import type { MockStore, MockStoreCreator } from 'redux-mock-store';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import {
  AnalystWorkspaceOperations,
  AnalystWorkspaceTypes
} from '../../../../src/ts/app/state/analyst';
import { analystSlice } from '../../../../src/ts/app/state/analyst/analyst-slice';
import { initialState } from '../../../../src/ts/app/state/reducer';
import type { AppState } from '../../../../src/ts/app/store';

const middlewares = [thunk];
const mockStoreCreator: MockStoreCreator<AppState, Redux.AnyAction> = createMockStore(middlewares);
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let store: MockStore<AppState, any>;

describe('state analyst-workspace operations', () => {
  describe('operations', () => {
    beforeEach(() => {
      store = mockStoreCreator({ app: initialState } as any);
    });

    describe('open event', () => {
      it('should set the open event (undefined) starting with initial state', () => {
        store.dispatch(AnalystWorkspaceOperations.setOpenEventId(undefined, undefined, undefined));
        const actions = store.getActions();
        expect(actions).toMatchSnapshot();
      });

      it('should set the open event (undefined)', () => {
        store = mockStoreCreator({
          app: {
            ...initialState,
            analyst: {
              ...initialState.analyst,
              selectedSdIds: ['selected-signal-detection-id'],
              openEventId: 'event',
              selectedEventIds: ['selected-event-id'],
              sdIdsToShowFk: ['signal-detection-id'],
              measurementMode: {
                mode: AnalystWorkspaceTypes.WaveformDisplayMode.MEASUREMENT,
                entries: { a: true }
              },
              selectedSortType: AnalystWorkspaceTypes.WaveformSortType.distance
            }
          }
        } as any);
        const expectedActions = [
          {
            type: analystSlice.actions.setOpenEventId.type,
            payload: undefined
          },
          {
            type: analystSlice.actions.setSelectedEventIds.type,
            payload: []
          },
          {
            type: analystSlice.actions.setSelectedSortType.type,
            payload: AnalystWorkspaceTypes.WaveformSortType.stationNameAZ
          },
          {
            type: analystSlice.actions.setMeasurementModeEntries.type,
            payload: {}
          },
          {
            type: analystSlice.actions.setSelectedLocationSolutionSetId.type,
            payload: undefined
          },
          {
            type: analystSlice.actions.setSelectedLocationSolutionId.type,
            payload: undefined
          },
          {
            type: analystSlice.actions.setSelectedPreferredLocationSolutionSetId.type,
            payload: undefined
          },
          {
            type: analystSlice.actions.setSelectedPreferredLocationSolutionId.type,
            payload: undefined
          },
          {
            type: analystSlice.actions.setSelectedSdIds.type,
            payload: []
          },
          {
            type: analystSlice.actions.setSdIdsToShowFk.type,
            payload: []
          },
          {
            type: analystSlice.actions.setMode.type,
            payload: AnalystWorkspaceTypes.WaveformDisplayMode.DEFAULT
          }
        ];
        store.dispatch(AnalystWorkspaceOperations.setOpenEventId(undefined, undefined, undefined));
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });
    });

    describe('measurement mode', () => {
      it('should set the mode to default', () => {
        const mode = AnalystWorkspaceTypes.WaveformDisplayMode.DEFAULT;
        const expectedActions = [{ type: analystSlice.actions.setMode.type, payload: mode }];

        store.dispatch(AnalystWorkspaceOperations.setMode(mode));
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });

      it('should set the mode to measurement', () => {
        const mode = AnalystWorkspaceTypes.WaveformDisplayMode.MEASUREMENT;
        const expectedActions = [{ type: analystSlice.actions.setMode.type, payload: mode }];

        store.dispatch(AnalystWorkspaceOperations.setMode(mode));
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });

      it('should set the measurements entries', () => {
        const entries = { a: true };
        const expectedActions = [
          {
            type: analystSlice.actions.setMeasurementModeEntries.type,
            payload: entries
          }
        ];

        store.dispatch(AnalystWorkspaceOperations.setMeasurementModeEntries(entries));
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });
    });

    describe('location', () => {
      it('should set the location solution', () => {
        const locationSolutionSetId = 'location-solution-set-id';
        const locationSolutionId = 'location-solution-id';
        const expectedActions = [
          {
            type: analystSlice.actions.setSelectedLocationSolutionSetId.type,
            payload: locationSolutionSetId
          },
          {
            type: analystSlice.actions.setSelectedLocationSolutionId.type,
            payload: locationSolutionId
          }
        ];
        store.dispatch(
          AnalystWorkspaceOperations.setSelectedLocationSolution(
            locationSolutionSetId,
            locationSolutionId
          )
        );
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });

      it('should set the preferred location solution', () => {
        const preferredLocationSolutionSetId = 'preferred-location-solution-set-id';
        const preferredLocationSolutionId = 'preferred-location-solution-id';
        const expectedActions = [
          {
            type: analystSlice.actions.setSelectedPreferredLocationSolutionSetId.type,
            payload: preferredLocationSolutionSetId
          },
          {
            type: analystSlice.actions.setSelectedPreferredLocationSolutionId.type,
            payload: preferredLocationSolutionId
          }
        ];
        store.dispatch(
          AnalystWorkspaceOperations.setSelectedPreferredLocationSolution(
            preferredLocationSolutionSetId,
            preferredLocationSolutionId
          )
        );
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
      });
    });
  });
});
