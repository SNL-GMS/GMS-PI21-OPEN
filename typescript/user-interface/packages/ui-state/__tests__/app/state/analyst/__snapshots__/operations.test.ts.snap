// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state analyst-workspace operations operations open event should set the open event (undefined) starting with initial state 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "analyst/setOpenEventId",
  },
  Object {
    "payload": Array [],
    "type": "analyst/setSelectedEventIds",
  },
  Object {
    "payload": "Station: A-Z",
    "type": "analyst/setSelectedSortType",
  },
  Object {
    "payload": Object {},
    "type": "analyst/setMeasurementModeEntries",
  },
  Object {
    "payload": undefined,
    "type": "analyst/setSelectedLocationSolutionSetId",
  },
  Object {
    "payload": undefined,
    "type": "analyst/setSelectedLocationSolutionId",
  },
  Object {
    "payload": undefined,
    "type": "analyst/setSelectedPreferredLocationSolutionSetId",
  },
  Object {
    "payload": undefined,
    "type": "analyst/setSelectedPreferredLocationSolutionId",
  },
  Object {
    "payload": Array [],
    "type": "analyst/setSelectedSdIds",
  },
  Object {
    "payload": Array [],
    "type": "analyst/setSdIdsToShowFk",
  },
  Object {
    "payload": "Default",
    "type": "analyst/setMode",
  },
]
`;
